bip-0331.mediawiki

```diff
@@ -0,0 +1,430 @@
+
+<br />
+
+'''No Package Information Round:''' One proposal is to just use the child's wtxid to refer to the
+package and always send the entire package together, skipping the package information round.
+However, this protocol would make it very likely for honest nodes to redownload duplicate
+transactions. See the following example, where the high-feerate ancestors were already downloaded
+and accepted individually.
+
+[[File:./bip-0331/no_package_info.png|600px]]
+<br />
+
+'''Package Information Only:''' Just having package information gives enough information for the
+receiver to accept the packages. Omit the "getpkgtxns" and "pkgtxns" messages and just download the
+transactions individually using existing messages. While this option is a potential fallback if
+batched transaction download fails for some reason, it shouldn't be used as the default because it
+always requires storage of unvalidated transactions.
+[[File:./bip-0331/package_info_only.png|1000px]]
+</ref>
+
+This protocol can be extended to include more types of package information in the future, while
+continuing to use the same messages for transaction data download.
+[[File:./bip-0331/generic_package_relay.png|600px]]
+
+Package relay is negotiated between two peers during the version handshake using a "sendpackages"
+message. The versions field within "sendpackages" is interpreted as a bitfield; peers may relay
+multiple versions of packages. Package relay requires both peers to support wtxid-based relay
+because package transactions are referenced by their wtxids.
+<ref>'''Why do we need multiple versions? Why can't we just support arbitrary packages?'''
+Attempting to support arbitrary packages in mempool validation may result in very complex logic, new
+Denial of Service attack vectors, and policy limitations that could be leveraged to censor
+transactions (aka "pinning attacks"). This protocol is extensible to support up to 32 types of
+packages based on future desired use cases. Future package information messages may describe
+different types of packages and/or contain more information than a list of wtxids, e.g. feerate or
+relationships between transactions.</ref>
+<ref>'''Why send one message per version instead of sending a single message for the highest
+supported version?''' It should also be possible to support some subset of the existing package
+types. For example, if a node's mempool policy doesn't support or a node implementation stops
+supporting version n, it should be easy to announce support for version n-1, but not version n,
+packages.</ref>
+
+[[File:./bip-0331/version_negotiation.png|400px]]
+
+===Ancestor Package Information===
+
+Ancestor Package Relay allows nodes to send and request a transaction's list of unconfirmed
+ancestors.  Nodes indicate support for Ancestor Package Relay by adding the
+<code>PKG_RELAY_ANCPKG = (1 << 0)</code> bit in their "sendpackages" messages during version handshake.
+
+===Intended Protocol Flow===
```

Comment:
While this section does talk about protocol flow it does so in the context of specific examples. I would probably call this section "Practical application of the Protocol with examples" or something like that, and move it out of and right after the specification section. I.e. have it between Specification and Compatibility. But I don't feel too strongly about this.

Reply:
I've changed this to "Protocol Flow Examples" but kept them in the same place. I'm thinking they might be helpful earlier on as a quick way to learn how the dialogue looks. Lmk if you disagree though!